### PR TITLE
Saner recursive lunit mock

### PR DESCRIPTION
**Pull request by halfflat**
* Split lunit.mock_function into (global) environment mocking and upvalue mocking functions.
* Allow recursive mocking to be optional and implement recursive mocking for upvalues.
* Add lunit unit tests including specifically tests for recursive mocking.
* Add lunit unit test run to Makefile target.

### CODE REVIEW

File: unit/test_cli_filter.lua
Comment (pelahi) [inline unit/test_cli_filter.lua:1]: Could you add some comments to the top of the file summarising the test?

File: unit/test_lunit.lua
Comment (pelahi) [inline unit/test_lunit.lua:1]: Similar comment, could you add some comments at the top summarising the test?

**Review by pelahi — COMMENTED** [review]
The tests look fine though both tests could do with some comments about what they test and what a failure might indicate. 
The refactor of the functions in lunit seem fine though I haven't tried them yet.

**Comment by pelahi** [comment]
Another quick thought regarding the tests. I haven't tried in lua but could the test, when run report the git commit hash? If we are dealing with issues in the filter, this would make it easier to track if there is a regression.

**Comment by halfflat** [comment]
Regarding git hash: thinking just reporting in unit test output preamble, or also from the filter itself as debug info?

File: unit/test_cli_filter.lua
Comment (halfflat) [reply unit/test_cli_filter.lua:1]: Added a top comment; is this the right sort of thing?

File: unit/test_lunit.lua
Comment (halfflat) [reply unit/test_lunit.lua:1]: Added top comment here too.

File: unit/test_cli_filter.lua
Comment (pelahi) [reply unit/test_cli_filter.lua:1]: Yup

**Review by pelahi — APPROVED** [review]

**Comment by halfflat** [comment]
* Split lunit.mock_function into (global) environment mocking and upvalue mocking functions.
* Allow recursive mocking to be optional and implement recursive mocking for upvalues.
* Add lunit unit tests including specifically tests for recursive mocking.
* Add lunit unit test run to Makefile target.

**Pull request merged**